File: packages/go/migrate.go
package alexsdk

// MigrateManifest upgrades a v1 manifest map to the EE-canonical v2 taxonomy.
// Returns (manifest, warnings, errors). If errors is non-empty, migration
// cannot proceed (un-migratable kind). Warnings are non-fatal issues.
//
// Kind remap:  tool -> mcp|atool (by transport: grpc=>atool, else mcp);
//
//	skill -> aagent;  agent -> aagent;  bundle -> aagent.
//
// Field remap: model stays model (EE uses `model`); llm/model_hint -> model;
//
//	aagent tags dropped (no such field in EE AagentConfig).
//
// Callers should write the returned manifest to atool.json after migration.
func MigrateManifest(v1 map[string]interface{}) (manifest map[string]interface{}, warnings []string, errors []string) {
	m := shallowCopy(v1)

	// Bump schema_version
	m["schema_version"] = "2"

	kind, _ := m["kind"].(string)

	// Handle removed kinds
	if kind == "llm-runtime" || kind == "llm-backend" {
		errors = append(errors,
			"kind '"+kind+"' has no v2 equivalent; register a model via "+
				"`alexandria install <name> --model` (.amodel) instead",
		)
		return m, warnings, errors
	}

	switch kind {
	case "bundle":
		// A bundle collapses to an aagent orchestrator carrying components[] refs.
		m["kind"] = "aagent"
		warnings = append(warnings, "bundle converted to aagent; add config.system_prompt before publishing")
		if cfg, ok := m["config"].(map[string]interface{}); ok {
			if oldComponents, ok := cfg["components"].([]interface{}); ok {
				newComps := make([]interface{}, 0, len(oldComponents))
				for _, ref := range oldComponents {
					newComps = append(newComps, map[string]interface{}{"ref": ref})
				}
				m["components"] = newComps
			}
		}
		m["config"] = map[string]interface{}{
			"kind":          "aagent",
			"system_prompt": "TODO: add system_prompt",
		}
	case "tool":
		// A v1 tool becomes mcp (MCP JSON-RPC/SSE) or atool (native gRPC),
		// discriminated by transport. Default (no transport) is MCP over http.
		cfg, _ := m["config"].(map[string]interface{})
		transport := ""
		if cfg != nil {
			transport, _ = cfg["transport"].(string)
		}
		newKind := "mcp"
		if transport == "grpc" {
			newKind = "atool"
		}
		m["kind"] = newKind
		if cfg != nil {
			cfg["kind"] = newKind
			m["config"] = cfg
		}
		if newKind == "atool" {
			warnings = append(warnings, "kind 'tool' with transport=grpc migrated to kind 'atool' (native ToolService)")
		} else {
			warnings = append(warnings, "kind 'tool' migrated to kind 'mcp' (MCP JSON-RPC/SSE)")
		}
	case "skill":
		// EE has no standalone skill kind — a skill is reusable prompt text that
		// ships as an aagent whose content is its system_prompt.
		m["kind"] = "aagent"
		if cfg, ok := m["config"].(map[string]interface{}); ok {
			cfg["kind"] = "aagent"
			m["config"] = cfg
		}
		warnings = append(warnings, "kind 'skill' migrated to kind 'aagent' (skills live in aagent.system_prompt)")
	case "agent":
		m["kind"] = "aagent"
		if cfg, ok := m["config"].(map[string]interface{}); ok {
			cfg["kind"] = "aagent"
			m["config"] = cfg
		}
	}

	// Migrate config fields to EE serde names.
	if cfg, ok := m["config"].(map[string]interface{}); ok {
		// EE uses `model`; the intermediate SDK-v2 field `llm` folds back to it.
		if llm, ok := cfg["llm"]; ok {
			cfg["model"] = llm
			delete(cfg, "llm")
			warnings = append(warnings, "config.llm renamed to config.model")
		}
		if modelHint, ok := cfg["model_hint"]; ok {
			cfg["model"] = modelHint
			delete(cfg, "model_hint")
			warnings = append(warnings, "config.model_hint renamed to config.model")
		}
		if _, ok := cfg["default_mode"]; ok {
			delete(cfg, "default_mode")
			warnings = append(warnings, "config.default_mode removed (swarm is always default)")
		}
		if _, ok := cfg["tags"]; ok {
			delete(cfg, "tags")
			warnings = append(warnings, "config.tags removed (EE aagent has no tags field)")
		}
		m["config"] = cfg
	}

	// Strip old signing fields at wrong locations
	var strippedSigning []string
	for _, field := range []string{"signed_at", "key_fingerprint"} {
		if _, ok := m[field]; ok {
			delete(m, field)
			strippedSigning = append(strippedSigning, field)
		}
	}
	// If signature present but not in v2 shape, remove it
	if sig, ok := m["signature"]; ok {
		sigMap, isMap := sig.(map[string]interface{})
		hasV2Shape := isMap &&
			sigMap["alg"] != nil &&
			sigMap["key_fingerprint"] != nil &&
			sigMap["value"] != nil &&
			sigMap["scope"] != nil
		if !hasV2Shape {
			delete(m, "signature")
			strippedSigning = append(strippedSigning, "signature")
		}
	}
	if len(strippedSigning) > 0 {
		msg := "signing fields removed ("
		for i, f := range strippedSigning {
			if i > 0 {
				msg += ", "
			}
			msg += f
		}
		msg += "); re-sign after migration"
		warnings = append(warnings, msg)
	}

	// Warn about dependencies missing version
	if deps, ok := m["dependencies"].([]interface{}); ok {
		for _, dep := range deps {
			if depMap, ok := dep.(map[string]interface{}); ok {
				if depMap["version"] == nil || depMap["version"] == "" {
					name, _ := depMap["name"].(string)
					if name == "" {
						name = "?"
					}
					warnings = append(warnings, "dependency '"+name+"' missing version field; add before publishing")
				}
			}
		}
	}

	return m, warnings, errors
}

func shallowCopy(m map[string]interface{}) map[string]interface{} {
	out := make(map[string]interface{}, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

File: packages/rust/alex-sdk/src/migrate.rs
//! Migration from v1 to v2 atool manifests.
//!
//! `migrate_manifest` takes a parsed v1 JSON value and returns a v2 JSON value
//! plus lists of warnings and errors. Callers write the result to `atool.json`.

use serde_json::{Map, Value};

/// Result of a migration pass.
pub struct MigrateResult {
    /// The (possibly transformed) manifest value.
    pub manifest: Value,
    /// Non-fatal issues — things that were changed but may require human review.
    pub warnings: Vec<String>,
    /// Fatal issues — the manifest cannot be migrated. If non-empty, the
    /// caller should abort and report errors without writing the output.
    pub errors: Vec<String>,
}

/// Migrate a v1 manifest JSON value to v2.
///
/// The input is consumed and transformed in-place where possible. Returns a
/// [`MigrateResult`] whose `manifest` field is the upgraded value.
pub fn migrate_manifest(mut v1: Value) -> MigrateResult {
    let mut warnings: Vec<String> = Vec::new();
    let mut errors: Vec<String> = Vec::new();

    let obj = match v1.as_object_mut() {
        Some(o) => o,
        None => {
            errors.push("manifest is not a JSON object".to_string());
            return MigrateResult { manifest: v1, warnings, errors };
        }
    };

    // Bump schema_version
    obj.insert("schema_version".to_string(), Value::String("2".to_string()));

    let kind = obj
        .get("kind")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    // Handle removed kinds
    if kind == "llm-runtime" || kind == "llm-backend" {
        errors.push(format!(
            "kind '{}' has no v2 equivalent; register via `alexandria llm install` instead",
            kind
        ));
        return MigrateResult {
            manifest: Value::Object(obj.clone()),
            warnings,
            errors,
        };
    }

    if kind == "bundle" {
        obj.insert("kind".to_string(), Value::String("agent".to_string()));
        warnings.push("bundle converted to agent; add config.system_prompt before publishing".to_string());

        // Convert bundleConfig.components -> top-level components[]
        let old_comps: Vec<Value> = if let Some(cfg) = obj.get("config").and_then(|c| c.as_object()) {
            cfg.get("components")
                .and_then(|c| c.as_array())
                .cloned()
                .unwrap_or_default()
        } else {
            Vec::new()
        };
        let new_comps: Vec<Value> = old_comps
            .into_iter()
            .map(|ref_val| {
                let mut m = Map::new();
                m.insert("ref".to_string(), ref_val);
                Value::Object(m)
            })
            .collect();
        obj.insert("components".to_string(), Value::Array(new_comps));

        // Replace bundle config with minimal agent config
        let mut new_cfg = Map::new();
        new_cfg.insert("kind".to_string(), Value::String("agent".to_string()));
        new_cfg.insert(
            "system_prompt".to_string(),
            Value::String("TODO: add system_prompt".to_string()),
        );
        obj.insert("config".to_string(), Value::Object(new_cfg));
    }

    // Migrate config fields
    if let Some(cfg) = obj.get_mut("config").and_then(|c| c.as_object_mut()) {
        if let Some(model) = cfg.remove("model") {
            cfg.insert("llm".to_string(), model);
            warnings.push("config.model renamed to config.llm".to_string());
        }
        if let Some(model_hint) = cfg.remove("model_hint") {
            cfg.insert("llm".to_string(), model_hint);
            warnings.push("config.model_hint renamed to config.llm".to_string());
        }
        if cfg.remove("default_mode").is_some() {
            warnings.push("config.default_mode removed (swarm is always default)".to_string());
        }
        // Warn about default_port: 0
        if let Some(dp) = cfg.get("default_port") {
            let is_zero = dp.as_u64().map(|v| v == 0).unwrap_or(false)
                || dp.as_f64().map(|v| v == 0.0).unwrap_or(false);
            if is_zero {
                warnings.push(
                    "default_port was 0 (schema-invalid); set to a valid port 1-65535".to_string(),
                );
            }
        }
    }

    // Strip old signing fields at wrong locations
    let mut stripped_signing: Vec<String> = Vec::new();
    for field in &["signed_at", "key_fingerprint"] {
        if obj.remove(*field).is_some() {
            stripped_signing.push(field.to_string());
        }
    }
    // If signature present but not in v2 shape, remove it
    if let Some(sig) = obj.get("signature") {
        let has_v2_shape = sig.get("alg").is_some()
            && sig.get("key_fingerprint").is_some()
            && sig.get("value").is_some()
            && sig.get("scope").is_some();
        if !has_v2_shape {
            obj.remove("signature");
            stripped_signing.push("signature".to_string());
        }
    }
    if !stripped_signing.is_empty() {
        warnings.push(format!(
            "signing fields removed ({}); re-sign after migration",
            stripped_signing.join(", ")
        ));
    }

    // Warn about dependencies missing version
    if let Some(deps) = obj.get("dependencies").and_then(|d| d.as_array()) {
        for dep in deps {
            let version_missing = dep
                .get("version")
                .map(|v| v.is_null() || v.as_str().map(|s| s.is_empty()).unwrap_or(false))
                .unwrap_or(true);
            if version_missing {
                let name = dep
                    .get("name")
                    .and_then(|n| n.as_str())
                    .unwrap_or("?")
                    .to_string();
                warnings.push(format!(
                    "dependency '{}' missing version field; add before publishing",
                    name
                ));
            }
        }
    }

    MigrateResult {
        manifest: Value::Object(obj.clone()),
        warnings,
        errors,
    }
}
